packages/dd-trace/src/plugins/util/web.js

```diff
@@ -221,107 +247,85 @@ const web = {
       const context = contexts.get(req)
       context.error = context.error || error
     }
-  }
-}
-
-function startSpan (tracer, config, req, res, name) {
-  const context = contexts.get(req)
-
-  context.config = config
-
-  let span
-
-  if (context.span) {
-    context.span.context()._name = name
-    span = context.span
-  } else {
-    span = web.startChildSpan(tracer, name, req.headers)
-  }
-
-  context.tracer = tracer
-  context.span = span
-  context.res = res
-
-  return span
-}
+  },
 
-function finish (context) {
-  const { req, res } = context
+  finishMiddleware (context) {
+    if (context.finished) return
 
-  if (context.finished && !req.stream) return
+    let span
 
-  addRequestTags(context)
-  addResponseTags(context)
+    while ((span = context.middleware.pop())) {
+      span.finish()
+    }
+  },
 
-  context.config.hooks.request(context.span, req, res)
-  addResourceTag(context)
+  finishSpan (context) {
+    const { req, res } = context
 
-  context.span.finish()
-  context.finished = true
-}
+    if (context.finished && !req.stream) return
 
-function finishMiddleware (context) {
-  if (context.finished) return
+    addRequestTags(context)
+    addResponseTags(context)
 
-  let span
+    context.config.hooks.request(context.span, req, res)
+    addResourceTag(context)
 
-  while ((span = context.middleware.pop())) {
-    span.finish()
-  }
-}
+    context.span.finish()
+    context.finished = true
+  },
+  wrapWriteHead (context) {
+    const { req, res } = context
+    const writeHead = res.writeHead
 
-function wrapEnd (context) {
-  const scope = context.tracer.scope()
-  const req = context.req
-  const res = context.res
-  const end = res.end
+    return function (statusCode, statusMessage, headers) {
+      headers = typeof statusMessage === 'string' ? headers : statusMessage
+      headers = Object.assign(res.getHeaders(), headers)
 
-  res.writeHead = wrapWriteHead(context)
+      if (req.method.toLowerCase() === 'options' && isOriginAllowed(req, headers)) {
+        addAllowHeaders(req, res, headers)
+      }
 
-  ends.set(res, function () {
-    for (const beforeEnd of context.beforeEnd) {
-      beforeEnd()
+      return writeHead.apply(this, arguments)
     }
+  },
+  wrapEnd (context) {
```

Comment:
Why is this still needed? Also, it looks like it's called by the new plugin even though the events should be doing the same thing?